src/What4/Serialize/FastSExpr.hs

```diff
@@ -0,0 +1,132 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE OverloadedStrings #-}
+-- | This module implements a specialized s-expression parser
+--
+-- The parser in s-cargot is very general, but that also makes it a bit
+-- inefficient.  This module implements a drop-in replacement parser for the one
+-- in What4.Serialize.Parser using megaparsec.  It is completely specialized to
+-- the types in this library.
+module What4.Serialize.FastSExpr (
+  parseSExpr
+  ) where
+
+import           Control.Applicative
+import qualified Control.Monad.Fail as MF
+import           Data.Parameterized.Some ( Some(..) )
+import           Data.Ratio ( (%) )
+import qualified Data.SCargot.Repr.WellFormed as SC
+import qualified Data.Text as T
+import qualified Data.Void as DV
+import           Numeric.Natural ( Natural )
+import qualified Text.Megaparsec as TM
+import qualified Text.Megaparsec.Char as TMC
+import qualified Text.Megaparsec.Char.Lexer as TMCL
+import qualified What4.BaseTypes as WT
+import qualified What4.Serialize.SETokens as WST
+
+-- | Parse 'T.Text' into the well-formed s-expression type from s-cargot.
+parseSExpr :: T.Text -> Either String (SC.WellFormedSExpr WST.Atom)
+parseSExpr t =
+  case TM.runParser (ws >> parse) "<input>" t of
+    Left errBundle -> Left (show errBundle)
+    Right a -> Right a
+
+type Parser a = TM.Parsec DV.Void T.Text a
+
+parse :: Parser (SC.WellFormedSExpr WST.Atom)
+parse = parseList <|> (SC.WFSAtom <$> lexeme parseAtom)
+
+parseList :: Parser (SC.WellFormedSExpr WST.Atom)
+parseList = do
+  _ <- lexeme (TMC.char '(')
+  items <- TM.many parse
+  _ <- lexeme (TMC.char ')')
+  return (SC.WFSList items)
+
+parseId :: Parser T.Text
+parseId = T.pack <$> ((:) <$> first <*> TM.many rest)
+  where
+    w4symbol c = c == '@'
+               || c == '+'
+               || c == '-'
+               || c == '='
+               || c == '<'
+               || c == '>'
+               || c == '_'
+               || c == '.'
```

Comment:
More compact: ```w4symbol c = c `elem` ("@+-=<>_." :: String)```

Reply:
Megaparsec (in the haddocks for `oneOf` https://hackage.haskell.org/package/megaparsec-8.0.0/docs/Text-Megaparsec.html#v:oneOf) recommends the explicit `||` form for performance; presumably to encourage GHC to inline the small conditional

Reply:
Oh, good to know!